wgengine/wgcfg/parser_test.go

```diff
@@ -53,3 +58,47 @@ func TestParseEndpoint(t *testing.T) {
 		t.Error("Error was expected")
 	}
 }
+
+func BenchmarkFromUAPI(b *testing.B) {
+	newPrivateKey := func() (wgkey.Key, wgkey.Private) {
+		b.Helper()
+		pk, err := wgkey.NewPrivate()
+		if err != nil {
+			b.Fatal(err)
+		}
+		return wgkey.Key(pk.Public()), wgkey.Private(pk)
+	}
+	k1, pk1 := newPrivateKey()
+	ip1 := netaddr.MustParseIPPrefix("10.0.0.1/32")
+
+	cfg1 := &Config{
+		PrivateKey: wgkey.Private(pk1),
+		Peers: []Peer{{
+			PublicKey:  k1,
+			AllowedIPs: []netaddr.IPPrefix{ip1},
+		}, {
+			PublicKey:  k1,
+			AllowedIPs: []netaddr.IPPrefix{ip1},
+		}, {
+			PublicKey:  k1,
+			AllowedIPs: []netaddr.IPPrefix{ip1},
+		}, {
+			PublicKey:  k1,
+			AllowedIPs: []netaddr.IPPrefix{ip1},
+		}},
+	}
+
+	buf := new(bytes.Buffer)
+	w := bufio.NewWriter(buf)
+	if err := cfg1.ToUAPI(w, &Config{}); err != nil {
+		b.Fatal(err)
+	}
+	w.Flush()
+	b.ReportAllocs()
+	for i := 0; i < b.N; i++ {
+		_, err := FromUAPI(buf)
```

Comment:
you're not benchmarking what you think you are.

You're measuring an empty parser:

https://play.golang.org/p/8hxjn_F5Jaj

<img width="637" alt="Screen Shot 2021-07-12 at 10 10 53 PM" src="https://user-images.githubusercontent.com/2621/125394110-07aad800-e35e-11eb-955d-600c2379ec11.png">

Reply:
I'm pretty sure I have an actual cfg, I printed out the value

Reply:
edit3: 0% sure :)

To convince yourself, add:

```go
        b.ReportAllocs()
        for i := 0; i < b.N; i++ {
                if buf.Len() == 0 {
                        b.Fatal("empty")
                }
                _, err := FromUAPI(buf)
                if err != nil {
                        b.Errorf("failed from UAPI: %v", err)
                }
        }
```